File: src/bot/photoSearch.ts
import { Bot, Context } from "grammy";
import FormData from "form-data";

const WARNING_MSG =
  "⚠️ Бот использует только открытые источники. Продолжая, вы подтверждаете, что " +
  "имеете право искать объекты/людей на этом фото (исследование цифрового следа) и не нарушаете закон.";

const DISCLAIMER_MSG =
  "\n\nℹ️ <i>Мы не храним переданные фотографии (обработка в ОЗУ). Прямые ссылки сгенерированы для безопасного поиска через ваш браузер.</i>";

// Simple in-memory state (instead of full FSM plugin to keep dependencies light)
const userStates = new Map<number, string>();

async function waitAndGetPhoto(ctx: Context, fileId: string) {
    const file = await ctx.api.getFile(fileId);
    if (!file.file_path) throw new Error("File path not found");
    const downloadUrl = `https://api.telegram.org/file/bot${ctx.api.token}/${file.file_path}`;
    const res = await fetch(downloadUrl);
    return res.arrayBuffer();
}

async function uploadToTelegraph(imageBuffer: ArrayBuffer) {
  try {
    const form = new FormData();
    form.append("file", Buffer.from(imageBuffer), {
      filename: "image.jpg",
      contentType: "image/jpeg",
    });

    const response = await fetch("https://telegra.ph/upload", {
      method: "POST",
      body: form as any,
    });

    if (response.status === 200) {
      const data: any = await response.json();
      if (data && data[0] && data[0].src) {
        return `https://telegra.ph${data[0].src}`;
      }
    }
    return null;
  } catch (err) {
    console.error("Telegraph Upload Failed:", err);
    return null;
  }
}

export function setupPhotoSearch(bot: Bot) {
  bot.command("search_photo", async (ctx) => {
    await ctx.reply(
      "📸 Пожалуйста, отправьте фотографию (как сжатое фото), по которой нужно выполнить обратный поиск."
    );
    if (ctx.from?.id) {
      userStates.set(ctx.from.id, "waiting_for_photo");
    }
  });

  bot.on("message:photo", async (ctx) => {
    const userId = ctx.from?.id;
    if (!userId || userStates.get(userId) !== "waiting_for_photo") {
       return; 
    }
    
    // Clear state
    userStates.delete(userId);
    
    await ctx.reply(WARNING_MSG);
    const statusMsg = await ctx.reply("⏳ Загружаю фотографию во временное хранилище (Telegraph) для генерации ссылок...");

    try {
      const photos = ctx.message.photo;
      const largestPhoto = photos[photos.length - 1]; // get highest res photo
      
      const buffer = await waitAndGetPhoto(ctx, largestPhoto.file_id);
      const publicUrl = await uploadToTelegraph(buffer);

      let text = "";
      if (publicUrl) {
        const enc = encodeURIComponent(publicUrl);
        text = `✅ <b>Фотография успешно обработана</b>\n\n` +
               `Поисковики блокируют автоматические запросы от ботов (поэтому Яндекс выдавал ошибку). Бот сгенерировал прямые ссылки, откройте их в браузере:\n\n` +
               `🔗 <a href="https://lens.google.com/uploadbyurl?url=${enc}">Поиск в Google Lens</a>\n` +
               `🔗 <a href="https://yandex.ru/images/search?rpt=imageview&url=${enc}">Поиск в Яндекс.Картинках</a>`;
      } else {
        text = "❌ К сожалению, не удалось загрузить фото для обратного поиска.";
      }

      await ctx.api.editMessageText(ctx.chat.id, statusMsg.message_id, text + DISCLAIMER_MSG, {
        parse_mode: "HTML",
        link_preview_options: { is_disabled: true },
      });

    } catch (err: any) {
      console.error(err);
      await ctx.api.editMessageText(ctx.chat.id, statusMsg.message_id, `❌ Ошибка при обработке изображения.\nДетали: ${err.message}`);
    }
  });
}

File: osint_bot/photo_search.py
import io
import json
import logging
import aiohttp
from aiogram import Router, types, F
from aiogram.filters import Command
from aiogram.fsm.context import FSMContext
from aiogram.fsm.state import State, StatesGroup

router = Router()
logger = logging.getLogger(__name__)

class PhotoSearchState(StatesGroup):
    waiting_for_photo = State()

WARNING_MSG = (
    "⚠️ Бот использует только открытые источники (Google/Yandex). Продолжая, вы подтверждаете, что "
    "имеете право искать объекты/людей на этом фото (исследование цифрового следа) и не нарушаете закон."
)

DISCLAIMER_MSG = (
    "\n\nℹ️ <i>Поиск осуществляется через публичные поисковые системы алгоритмами Content-Based Image Retrieval. "
    "Мы не храним переданные фотографии (обработка в ОЗУ).</i>"
)

@router.message(Command(commands=["search_photo"]))
async def search_photo_start(message: types.Message, state: FSMContext):
    await message.answer("📸 Пожалуйста, отправьте фотографию (как сжатое фото или документ), по которой нужно выполнить обратный поиск.")
    await state.set_state(PhotoSearchState.waiting_for_photo)
    
@router.message(PhotoSearchState.waiting_for_photo, F.photo)
async def process_photo(message: types.Message, state: FSMContext):
    await state.clear()
    await message.answer(WARNING_MSG)
    
    status_msg = await message.answer("⏳ Загружаю фотографию в оперативную память и отправляю запрос в поисковик...")
    
    try:
        # Берем фото максимального разрешения и сохраняем в ОЗУ
        photo = message.photo[-1]
        photo_bytes_io = await message.bot.download(photo, destination=io.BytesIO())
        photo_bytes = photo_bytes_io.read()
        
        # Выполняем публичный обратный поиск
        search_url = await _yandex_image_search(photo_bytes)
        
        if search_url:
            text = (
                f"✅ <b>Обратный поиск по изображению завершен</b>\n\n"
                f"Ссылка на результаты в Яндекс.Картинках (найденные лица, веб-страницы, совпадения):\n"
                f"🔗 <a href='{search_url}'>Перейти к результатам поиска</a>\n\n"
                f"<i>Рекомендуется открыть ссылку с браузера для детального анализа найденных страниц.</i>"
            )
        else:
            text = "❌ К сожалению, поисковик отклонил запрос или не вернул результатов."
            
        await status_msg.edit_text(text + DISCLAIMER_MSG, parse_mode="HTML", disable_web_page_preview=True)

    except Exception as e:
        logger.error(f"Error processing photo: {e}")
        await status_msg.edit_text(f"❌ Ошибка при обработке изображения.\nДетали: {e}")

async def _yandex_image_search(image_bytes: bytes) -> str | None:
    """
    Отправляет изображение в Yandex Images (поиск по картинке)
    через публичный HTTP endpoint CBIR (Content-Based Image Retrieval).
    """
    try:
        url = (
            "https://yandex.ru/images/search?"
            "rpt=imageview&format=json&request"
            "=%7B%22blocks%22%3A%5B%7B%22block%22%3A%22b-page_type_search-by-image__link%22%7D%5D%7D"
        )
        
        form = aiohttp.FormData()
        form.add_field('upfile', image_bytes, filename='image.jpg', content_type='image/jpeg')
        
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
        }
        
        async with aiohttp.ClientSession(headers=headers) as session:
            async with session.post(url, data=form, timeout=15) as resp:
                if resp.status == 200:
                    data = await resp.json()
                    # Извлекаем CBIR-идентификатор загруженной картинки для получения ссылки на результаты
                    params = data.get("blocks", [{}])[0].get("params", {})
                    cbir_id = params.get("url")
                    
                    if cbir_id:
                        return f"https://yandex.ru/images/search?rpt=imageview&url={cbir_id}"
                
                logger.warning(f"Yandex ответил статусом: {resp.status}")
                return None
    except Exception as e:
        logger.error(f"Yandex request failed: {e}")
        return None
